### ISSUE TITLE
twinBasic handling of parentheses in call statements

### ISSUE THREAD

**Issue by bclothier** [WaynePhillipsEA/twinbasic#262]
**Describe the bug**
tB is able to handle calls to sub with or without parentheses. Including parentheses usually is a compiler error as shown in a VBIDE:
![image](https://user-images.githubusercontent.com/2367644/124357691-7e1f3b80-dbe2-11eb-8d51-9de2e5c36376.png)

The same works without problem in tB:
![image](https://user-images.githubusercontent.com/2367644/124357719-a4dd7200-dbe2-11eb-95f4-9c541829af12.png)

On the surface, this seems an improvement and I like it very much. However, there is a difference when we start dealing with objects, particularly with default members.

Back to VBIDE:
![image](https://user-images.githubusercontent.com/2367644/124357760-cfc7c600-dbe2-11eb-8d78-f6ddea7c0c65.png)

The version calling with parentheses is actually equivalent to doing `Application.Name`:
![image](https://user-images.githubusercontent.com/2367644/124357800-ff76ce00-dbe2-11eb-8d5b-e0771c25d4c0.png)

If we do the same thing in tB:
![image](https://user-images.githubusercontent.com/2367644/124357826-23d2aa80-dbe3-11eb-997a-86b2c16dafc7.png)

The output returns 2 `Application`, rather than a `String` and a `Application`. 

If I modify the code to explicitly include `.Name`, we do get the `String`:
![image](https://user-images.githubusercontent.com/2367644/124357927-8d52b900-dbe3-11eb-8351-167f33315b0a.png)

**To Reproduce**
1. Paste the following code in a tB project:
```
    Public Application As Excel.Application
	
    ' This project type is set to 'Standard EXE' in the Settings file, so you need a Main() subroutine to run when the EXE is started.
    Public Sub Main()

        Set Application = New Excel.Application
        Test
	
    End Sub
	
    Public Sub Foo(a As String, b As Long, c As Date)
        Debug.Print "Fooed"
    End Sub
	
    Public Sub DoIt()
        Foo("abc", 123, #12-Apr-2021#)
    End Sub

    Public Sub Test()
        WithParentheses
        WithoutParentheses
    End Sub
	
    Public Sub WithParentheses()
        WhatIsIt(Application.Name)
    End Sub
	
    Public Sub WithoutParentheses()
        WhatIsIt Application
    End Sub
	
    Public Sub WhatIsIt(TheThing As Variant)
        Debug.Print TypeName(TheThing)
    End Sub
```
2. Put cursor in the `Main`, press F5 to run the test.
3. Paste the functions except for `Main` in an Excel VBA project and run the `Test` routine.
4. Compare the difference in debug output.

**Expected behavior**
I'll be blunt, I actually prefer tB's behavior over VB*'s behavior as the whole thing with "sometimes can, sometimes can't" approach to handling parentheses is very inconsistent and annoying. I hate those implicit code that doesn't say what it's doing and I hope that with codelen feature, we will be able to highlight the implicit member access to the default member of the object so it's more obvious. But, tB was made with backward compatibility as the goal. For that reason, we need to be able to deal with junk code where a procedure is sometimes called with parentheses, sometimes not. 

One approach would be to take a surgical approach -- make it a compile error if it's a procedure with a single argument that invokes an object or a `Variant`, call it ambiguous reference. That way, tB can keep the improvement of `Foo("abc", 123, #12-Apr-2021#)` which would never have worked in VBIDE so we don't have worry about legacy code with that syntax. For a procedure that has 2 or more parameters, the presence of commas within the parentheses _ought_ make it obvious that those are arguments, not an expression. I do not think it is possible to write an expression using comma anywhere besides the special forms such as `Circle`, `PSet`, etc. 

The alternative is to follow VB*' hair-pulling insane "no parentheses for calls without returns but parentheses for calls with return" convention. If we are going to recruit new people, I suspect they will quite dislike this, particularly if coming from other languages where the use of parentheses is more consistent and sane.

Also, I can confirm that this works:
```
	Public Sub WithParentheses()
		WhatIsIt((Application))
	End Sub
```
which will then return a `String`. 

Some relevant MS-VBAL references:

[5.6.6 Parenthesized Expressions](https://docs.microsoft.com/en-us/openspecs/microsoft_general_purpose_programming_languages/ms-vbal/8c1b7427-670d-45e2-bc4b-39ab84ba5b20)
[5.6.2.2 Evaluation to a simple data value](https://docs.microsoft.com/en-us/openspecs/microsoft_general_purpose_programming_languages/ms-vbal/f1cc9a8d-e681-4e20-9c5e-e3385545440e), specifically this paragraph:
> If this class has a public default Property Get or function and this default member’s parameter list is compatible with an argument list containing 0 parameters, simple data value evaluation restarts as if this default member was the expression.

and further:

> If the source object has a public default Property Get or a public default function, and this default member’s parameter list is compatible with an argument list containing 0 parameters, the simple data value’s value is the result of evaluating this default member as a simple data value. 

[5.5.1.1 Static semantics (for Let-coercion](https://docs.microsoft.com/en-us/openspecs/microsoft_general_purpose_programming_languages/ms-vbal/4ef84713-33da-445b-942e-345982b3267d) which lists 'Any class which has no accessible default Property" as a invalid example of let-coerecion.

**Desktop (please complete the following information):**
 - OS: [Windows 10]
 - twinBASIC compiler version [0.10.3412]

**Comment by WaynePhillipsEA** [comment]
Indeed, Foo(A) is not being handled correctly in the parser here, so that most definitely needs to be fixed for backwards compatibility.   

The main question seems to be whether we continue to support Foo(A,B), where (assuming A as an object/variant) the default value member of A would NOT be invoked in this syntax.  Given that the default member has to be invoked for Foo(A), that just doesn't feel right for consistency.  

I guess we could restrict the Foo(A) default member evaluation only to Legacy-mode, and then in non-legacy mode have a warning or hint about the difference when Foo(A) is encountered with A as an object/variant.

Or we just disallow Foo(A,B) and stick with the VB behaviour.   Needs more thought...

**Comment by bclothier** [comment]
In the case of `Foo(A,B)`, it will depend on the data types we're coercing to. A `Sub Foo(String A, String B)` would mean that a call of `Foo(Application, Application)` would be implicitly evaluated as if it was `Foo(Application.Name, Application.Name)`. That's where the section on Let-coercion applies. However, with a `Sub Foo(Variant A, Variant B)`, the call of `Foo(Application, Application)` would be a compile error in VB*, so we'll always have `Foo Application, Application`, which will provide us the objects, not the default member's return. 

I believe it's only the `Sub Foo(A As Variant)` _and_ the call of `Foo(Application)` that is potentially ambiguous because in VB*, it'll be evaluated as `Foo Application.Name`, and only if that object had a default member defined (else it's a runtime error). 

Some more demo code:
```
Public Sub WhatIsIt2(ParamArray Args() As Variant)
    Dim Arg As Variant
    
    For Each Arg In Args
        Debug.Print TypeName(Arg)
    Next
End Sub

Public Sub IsItAString(Arg As String)
    Debug.Print VarType(Arg) = vbString
End Sub

Public Sub Test2()
    'Returns "Application"
    WhatIsIt2 Application
    
    'Returns "String"
    WhatIsIt2 (Application)
    
    'Returns "Application, Application"
    WhatIsIt2 Application, Application
    
    'Returns "String, String"
    WhatIsIt2 (Application), (Application)
    
    'Returns True
    IsItAString (Application)
    
    'Returns True
    IsItAString Application
    
    'Compile error
    'WhatIsIt2 (Application, Application)
End Sub
```

**Comment by bclothier** [comment]
Addendum: I asserted that the defining function would have to use `Variant` as the data type of the parameter for this behavior to become problematic; however, thinking about it, it might be also true for `Object`, with the caveat that the default member of the argument being passed in must return an object or throw a run-time error. So I think it has to handle the case for both `Variant` and `Object`.

**Comment by Kr00l** [comment]
My 2 cent would be to make behavior strictly as in VBx. I kinda like this norm "no parentheses for calls without returns but parentheses for calls with return". :-)
And the less complexity, differences between VBx and tB the better..

**Comment by bclothier** [comment]
Yet there are people who argue that they should keep using the keyword `Call` just so they can get parentheses on calls without return. I disagree with using the obsolete keyword but I can see why they would feel that way.

**Comment by WaynePhillipsEA** [comment]
It's a difficult, and somewhat controversial issue.  If we were able to support Foo(A, B) in a consistent way, i.e. the same behaviour as Foo(A), and without changing the backwards compatible Foo(A) behaviour, then I'd argue to support it. But we can't, and therefore for me backwards compatibility has to win out, as that's always been the highest priority for this project.

VB.NET is an interesting case study here because they require parenthesis on all calls, but then there's no default member evaluation ambiguity... and ultimately no proper backwards compatibility in part due to that.

**Comment by bclothier** [comment]
Just a thought.

In VBx, the expression `Foo(A)` will always get prettified into `Foo (A)`.

Therefore, the tB parser could treat `Foo(A)` differently from `Foo (A)`. That would then allow for fully compatible porting while keeping the gain of feature.

**Comment by WaynePhillipsEA** [comment]
> Just a thought.
> 
> In VBx, the expression `Foo(A)` will always get prettified into `Foo (A)`.
> 
> Therefore, the tB parser could treat `Foo(A)` differently from `Foo (A)`. That would then allow for fully compatible porting while keeping the gain of feature.

I think that's a workable idea.  Also, I would guess that most uses of `Foo (A)` aren't actually written that way for the benefit of the default member evaluation anyway, e.g. like `MsgBox ("blah")` , so tB could be clever and remove the space in those instances where the default member evaluation isn't applicable.   I would say this should all be fine as it will honour backwards compatibility whilst allowing for the new feature.

**Comment by WaynePhillipsEA** [comment]
A typo was stopping the default member being invoked here.  It was working in other cases like `Foo.Bar (obj)`

**Comment by WaynePhillipsEA** [comment]
This should now be fixed in v0.10.3423.  I haven't applied the auto-removal of a redundant space, in an effort to try to please everybody :)
